[Composable API]Common _State parent class for composable and wrapper FSDP

fegin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #89147

**Why this PR?**

For the composable APIs implementation, sometimes the internal APIs may not have the application (FSDP, DDP) root module but only the local module. One example is the state_dict/optimizer_state_dict implementation of FSDP. These APIs  are designed to start with the root module of the model. It is tricky for these APIs to tell whether a random submodule is managed by either DDP or FSDP. 

It will be useful to have APIs like:
`_get_module_state(module)`: return the composable state if this module is managed by composable API.
`_get_module_fsdp_state(module)`: return the FSDP state if this module is managed by FSDP.

**What does this PR propose?**
1. Make `_State` out of `_composable` module so that `FullyShardedDataParallel` can inherit from it.
2. A global `_module_state_mapping: Dict[nn.Module, _State]` that keeps the mapping of all submodules (not just root module) to the state.
3. Create `_get_module_state(module)` to look up `_module_state_mapping`.
4. Create `_get_module_fsdp_state(module)` that uses `_get_module_state(module)` to get the state then verifies if the state is `_FSDPState`.

fegin committed:
[FSDP][Draft] Common _State parent class for composable and wrapper FSDP

[ghstack-poisoned]

rohan-varma commented on torch/distributed/_composable_state.py:
```diff
@@ -0,0 +1,21 @@
+    global _module_state_mapping
+    if isinstance(module, _State):
+        return cast(_State, module)
+    else:
```
just for my understanding, this is the fork in the path for composable vs non-composable scenarios, right?

fegin replied:
Yes, correct.

awgu replied:
IIUC: 
- The `if` branch is for non-composable, and the `else` branch is for composable.
- It is not a hard requirement for `DistributedDataParallel` to inherit from some `_DDPState` that inherits from `_State` unless we want to unify the underlying composable/non-composable code like for FSDP.

fegin replied:
I agree it is not a hard requirement for `DistributedDataParallel` to inherit from `_DDPState` but it won't change anything. If `DistributedDataParallel`  does not unify the code like FSDP but simply branch the code, no issues like FSDP will occur during the shared code path since the code is not unified. The `_DDPState` version can still work for the composable API.

yhcharles replied:
The current implementation in `Contract` is having a State per module per API. This PR will change it to per module State shared by all APIs, is my understanding correct?
cc @mrshenli

awgu commented on torch/distributed/_composable/fully_shard.py:
```diff
@@ -45,6 +47,8 @@ def fully_shard(
     # Enforce the new auto wrap policy
     if policy is not None and not isinstance(policy, _FSDPPolicy):
         raise ValueError(f"Expects an `_FSDPPolicy` but got {policy}")
+    # This is a hack to make _FSDPState to be compatible with the existing contract.
```
Can you explain why `fully_shard.set_state_module, _FSDPState())` is considered a hack to make `_FSDPState` to be compatible with the contract?

fegin replied:
I don't like the current code flow that `contract` first creates a `_State` instance and then `fully_shard` creates another one and overwrite the existing one. I think a better way to do that is to that `contract` wrapper to accept a `_State` subclass and instantiate the class.

yhcharles replied:
Agree, it is true for the`replicate` API as well, it would be better to allow the API to pass in a State object

awgu reviewed (commented):
The design behind this PR looks good to me! I think `_get_module_fsdp_state()` can benefit the FSDP runtime logic as well.

fegin committed:
Update on "[Composable API][Draft] Common _State parent class for composable and wrapper FSDP"




[ghstack-poisoned]

fegin committed:
Update on "[Composable API][Draft] Common _State parent class for composable and wrapper FSDP"




[ghstack-poisoned]

fegin committed:
Update on "[Composable API]Common _State parent class for composable and wrapper FSDP"




[ghstack-poisoned]

fegin committed:
Update on "[Composable API]Common _State parent class for composable and wrapper FSDP"




[ghstack-poisoned]

yhcharles commented on torch/distributed/_composable_state.py:
```diff
@@ -0,0 +1,21 @@
+
+def _get_module_state(module: nn.Module) -> Optional[_State]:
+    global _module_state_mapping
+    if isinstance(module, _State):
```
- Why `module` can be a `_State` instance?
- If it's necessary, then do we need to update the type of the `module` arg?

fegin replied:
This is actually for non-composable API, wrapper-based FSDP. FSDP shares most utils between composable and wrapper-based implementations. The reason we can do this check (specifically for wrapper-based FSDP), is that we add `class FullyShardedDataParallel(nn.Module, _FSDPState):` to ensure `_get_module_state` can correctly detect it. That's also the reason why `_get_module_state` is not inside `_composable/contract.py` -- to avoid circular dependency.

awgu commented on torch/distributed/_composable/fully_shard.py:
```diff
@@ -81,4 +83,6 @@ def fully_shard(
     _register_pre_forward_hooks(state, modules)
     _register_post_forward_hooks(state, modules)
     _register_root_pre_forward_hook(state, module)  # prepend last
+    for submodule in state._module_to_handles.keys():
```
I am not sure if this `for` loop is exactly what we want. This is probably my fault for making `_module_to_handles` confusing. The following is going to get into some details about FSDP.

The keys of the `dict` are computed as the union over all `handle._modules` for all `handles` in `state._handles`. However, `handle._modules` only includes the modules that have parameters in the `handle.flat_param`. 

For example, if you have
```
parent(
    child()
    child()
)
``` 
where only each `child` has parameters but `parent` does not, then `parent` will not be in `handle._modules`. In that case, IIUC, `parent` is not given access to the `state`. In other words, the invariant here is that, if we apply `fully_shard(module)`, then a submodule in `module`'s tree is given access to `state` iff the submodule has parameters.

Perhaps, this is _sufficient_ for your use case in checkpointing? Regardless, I can see that this might be confusing.

One way to fix this depends on one of my not-yet-landed PRs: https://github.com/pytorch/pytorch/pull/90387. That PR will introduce a `comm_module` for each `FlatParamHandle`, which for the wrapper code path is exactly the same as the `module` passed to `FlatParamHandle.__init__()` and for the composable path represents that same module. The difference is that for the composable `fully_shard` path, we always pass the local FSDP root module as `module` to `FlatParamHandle.__init__()`, which may not be the same module we would have passed into `__init__()` for the wrapper path. `comm_module` handles that difference.

If we have `comm_module`, then we can do something like:
```
all_submodules = set()
for handle in _all_handles(state):
    for submodule in handle._comm_module.modules():
        all_submodules.add(submodule)
for submodule in all_submodules:
    _insert_module_state(submodule, state)
```
Note that the above snippet has a quadratic complexity with the number of submodules, which is not ideal, but should be okay for now. The reason is that `_comm_module.modules()` will "double-count" modules if one `_comm_module` is the parent of another.

If the way you have it in this PR is okay for now, then we do not need to change this one. We can wait for my `comm_module` PR to land and then adjust after.

This is kind of convoluted, so please let me know if anything is unclear!

awgu replied:
Discussed offline: We can just do something like:
```
for submodule in module.modules():
    if submodule not in state._ignored_modules:
        _insert_module_state(submodule, state)
```
(without relying on `_module_to_handles`)

fegin committed:
Update on "[Composable API]Common _State parent class for composable and wrapper FSDP"


**Why this PR?**

For the composable APIs implementation, sometimes the internal APIs may not have the application (FSDP, DDP) root module but only the local module. One example is the state_dict/optimizer_state_dict implementation of FSDP. These APIs  are designed to start with the root module of the model. It is tricky for these APIs to tell whether a random submodule is managed by either DDP or FSDP. 

It will be useful to have APIs like:
`_get_module_state(module)`: return the composable state if this module is managed by composable API.
`_get_module_fsdp_state(module)`: return the FSDP state if this module is managed by FSDP.

**What does this PR propose?**
1. Make `_State` out of `_composable` module so that `FullyShardedDataParallel` can inherit from it.
2. A global `_module_state_mapping: Dict[nn.Module, _State]` that keeps the mapping of all submodules (not just root module) to the state.
3. Create `_get_module_state(module)` to look up `_module_state_mapping`.
4. Create `_get_module_fsdp_state(module)` that uses `_get_module_state(module)` to get the state then verifies if the state is `_FSDPState`.



[ghstack-poisoned]

yhcharles reviewed (dismissed):
Thanks @fegin, LGTM

awgu commented on torch/distributed/fsdp/_common_utils.py:
```diff
@@ -4,25 +4,38 @@
-_FSDPState = Union[nn.Module, Any]
+class _FSDPState(_State):
+    def __init__(self) -> None:
+        # Move all the attributes to this class to enable typing for FSDP/fully_shard.
```
This is awesome. I think it would be great to get type checking formalized for FSDP. I will need to follow-up to see if there are any cases that require some redefinition of data structures.

yhcharles commented on torch/distributed/_composable/contract.py:
```diff
@@ -57,96 +54,93 @@ def contract(func):
+        def wrapper(module: nn.Module, *args, **kwargs) -> Optional[nn.Module]:
+            # install states specific to the wrapped ``func``
+            default_all_state: Dict[Callable, _State] = {}
+            all_state: Dict[Callable, _State] = module.__dict__.setdefault(  # type: ignore[call-overload]
```
Should we update this mapping here to use the one in `_composable_state.py`? So that we are consistently using `module -> State` mapping.

fegin commented on torch/distributed/_composable/contract.py:
```diff
@@ -57,96 +54,93 @@ def contract(func):
-    wrapper.state = get_state  # type: ignore[attr-defined]
-
-    return wrapper
+    def inner(func):
```
This is like Python's rule to do decorator with argument. The outer function, `contract()` now accepts arguments. As a result, it cannot accepts func as an argument anymore. As a result, it has to return another decorator, `inner()` to accept `func` argument.

fegin committed:
Update on "[Composable API]Common _State parent class for composable and wrapper FSDP"


**Why this PR?**

For the composable APIs implementation, sometimes the internal APIs may not have the application (FSDP, DDP) root module but only the local module. One example is the state_dict/optimizer_state_dict implementation of FSDP. These APIs  are designed to start with the root module of the model. It is tricky for these APIs to tell whether a random submodule is managed by either DDP or FSDP. 

It will be useful to have APIs like:
`_get_module_state(module)`: return the composable state if this module is managed by composable API.
`_get_module_fsdp_state(module)`: return the FSDP state if this module is managed by FSDP.

**What does this PR propose?**
1. Make `_State` out of `_composable` module so that `FullyShardedDataParallel` can inherit from it.
2. A global `_module_state_mapping: Dict[nn.Module, _State]` that keeps the mapping of all submodules (not just root module) to the state.
3. Create `_get_module_state(module)` to look up `_module_state_mapping`.
4. Create `_get_module_fsdp_state(module)` that uses `_get_module_state(module)` to get the state then verifies if the state is `_FSDPState`.



[ghstack-poisoned]

fegin committed:
Update on "[Composable API]Common _State parent class for composable and wrapper FSDP"


**Why this PR?**

For the composable APIs implementation, sometimes the internal APIs may not have the application (FSDP, DDP) root module but only the local module. One example is the state_dict/optimizer_state_dict implementation of FSDP. These APIs  are designed to start with the root module of the model. It is tricky for these APIs to tell whether a random submodule is managed by either DDP or FSDP. 

It will be useful to have APIs like:
`_get_module_state(module)`: return the composable state if this module is managed by composable API.
`_get_module_fsdp_state(module)`: return the FSDP state if this module is managed by FSDP.

**What does this PR propose?**
1. Make `_State` out of `_composable` module so that `FullyShardedDataParallel` can inherit from it.
2. A global `_module_state_mapping: Dict[nn.Module, _State]` that keeps the mapping of all submodules (not just root module) to the state.
3. Create `_get_module_state(module)` to look up `_module_state_mapping`.
4. Create `_get_module_fsdp_state(module)` that uses `_get_module_state(module)` to get the state then verifies if the state is `_FSDPState`.



[ghstack-poisoned]

awgu reviewed (approved):
LGTM!

awgu commented:
Lint error is real.

fegin committed:
Update on "[Composable API]Common _State parent class for composable and wrapper FSDP"


**Why this PR?**

For the composable APIs implementation, sometimes the internal APIs may not have the application (FSDP, DDP) root module but only the local module. One example is the state_dict/optimizer_state_dict implementation of FSDP. These APIs  are designed to start with the root module of the model. It is tricky for these APIs to tell whether a random submodule is managed by either DDP or FSDP. 

It will be useful to have APIs like:
`_get_module_state(module)`: return the composable state if this module is managed by composable API.
`_get_module_fsdp_state(module)`: return the FSDP state if this module is managed by FSDP.

**What does this PR propose?**
1. Make `_State` out of `_composable` module so that `FullyShardedDataParallel` can inherit from it.
2. A global `_module_state_mapping: Dict[nn.Module, _State]` that keeps the mapping of all submodules (not just root module) to the state.
3. Create `_get_module_state(module)` to look up `_module_state_mapping`.
4. Create `_get_module_fsdp_state(module)` that uses `_get_module_state(module)` to get the state then verifies if the state is `_FSDPState`.



[ghstack-poisoned]

fegin committed:
Update on "[Composable API]Common _State parent class for composable and wrapper FSDP"


**Why this PR?**

For the composable APIs implementation, sometimes the internal APIs may not have the application (FSDP, DDP) root module but only the local module. One example is the state_dict/optimizer_state_dict implementation of FSDP. These APIs  are designed to start with the root module of the model. It is tricky for these APIs to tell whether a random submodule is managed by either DDP or FSDP. 

It will be useful to have APIs like:
`_get_module_state(module)`: return the composable state if this module is managed by composable API.
`_get_module_fsdp_state(module)`: return the FSDP state if this module is managed by FSDP.

**What does this PR propose?**
1. Make `_State` out of `_composable` module so that `FullyShardedDataParallel` can inherit from it.
2. A global `_module_state_mapping: Dict[nn.Module, _State]` that keeps the mapping of all submodules (not just root module) to the state.
3. Create `_get_module_state(module)` to look up `_module_state_mapping`.
4. Create `_get_module_fsdp_state(module)` that uses `_get_module_state(module)` to get the state then verifies if the state is `_FSDPState`.



[ghstack-poisoned]

awgu commented:
Test failure is real:
```
======================================================================
ERROR [4.213s]: test_float16_on_one_submodule (__main__.TestMixedPrecision)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/conda/lib/python3.10/site-packages/torch/testing/_internal/common_distributed.py", line 534, in wrapper
    self._join_processes(fn)
  File "/opt/conda/lib/python3.10/site-packages/torch/testing/_internal/common_distributed.py", line 760, in _join_processes
    self._check_return_codes(elapsed_time)
  File "/opt/conda/lib/python3.10/site-packages/torch/testing/_internal/common_distributed.py", line 805, in _check_return_codes
    raise RuntimeError(error)
RuntimeError: Process 0 exited with error code 10 and exception:
Traceback (most recent call last):
  File "/opt/conda/lib/python3.10/site-packages/torch/testing/_internal/common_distributed.py", line 658, in run_test
    getattr(self, test_name)()
  File "/opt/conda/lib/python3.10/site-packages/torch/testing/_internal/common_distributed.py", line 536, in wrapper
    fn()
  File "/opt/conda/lib/python3.10/site-packages/torch/testing/_internal/common_distributed.py", line 167, in wrapper
    return func(*args, **kwargs)
  File "/var/lib/jenkins/workspace/test/distributed/_composable/test_fully_shard.py", line 433, in test_float16_on_one_submodule
    fsdp = fully_shard(model)
  File "/opt/conda/lib/python3.10/site-packages/torch/distributed/_composable/contract.py", line 107, in wrapper
    updated = func(module, *args, **kwargs)
  File "/opt/conda/lib/python3.10/site-packages/torch/distributed/_composable/fully_shard.py", line 93, in fully_shard
    _insert_module_state(submodule, state)
  File "/opt/conda/lib/python3.10/site-packages/torch/distributed/_composable_state.py", line 15, in _insert_module_state
    assert module not in _module_state_mapping, f"Inserting {module} more than once."
AssertionError: Inserting SaveForwardInputsModule(
  (l): Linear(in_features=100, out_features=100, bias=True)
) more than once.
```

awgu commented on torch/distributed/_composable/fully_shard.py:
```diff
@@ -81,4 +85,10 @@ def fully_shard(
     _register_root_pre_forward_hook(state, module)  # prepend last
+    for submodule in module.modules():
+        if (
+            state._ignored_modules is not None
```
nit: Do we need this `is not None` check? IIUC, `state._ignored_modules` should not be `None`.

awgu commented on torch/distributed/_composable_state.py:
```diff
@@ -0,0 +1,24 @@
+    _module_state_mapping[module] = state
+
+
+def _get_module_state(module: nn.Module) -> Optional[_State]:
+    global _module_state_mapping
+    if isinstance(module, _State):
+        return cast(_State, module)
+    else:
+        return _module_state_mapping.get(module, None)
```
nit: Do you think this might be clearer? (up to you)
```suggestion
def _get_module_state(module_or_state: Union[nn.Module, _State]) -> Optional[_State]:
    global _module_state_mapping
    if isinstance(module_or_state, _State):
        return cast(_State, module_or_state)
    else:
        return _module_state_mapping.get(module_or_state, None)
```

fegin replied:
This is semantically not correct. The API is used to map from `module` to `state` where `state` can be either a module (module-wrapped case) or `_FSDPSTate`. So the input argument must be a module.

awgu replied:
Ohh ok. I understand what is going on now. `FullyShardedDataParallel` inherits from `_FSDPState`, which inherits from `_State`, so wrapper FSDP is the `if isinstance(module, _State)` code path.

fegin committed:
Update on "[Composable API]Common _State parent class for composable and wrapper FSDP"


**Why this PR?**

For the composable APIs implementation, sometimes the internal APIs may not have the application (FSDP, DDP) root module but only the local module. One example is the state_dict/optimizer_state_dict implementation of FSDP. These APIs  are designed to start with the root module of the model. It is tricky for these APIs to tell whether a random submodule is managed by either DDP or FSDP. 

It will be useful to have APIs like:
`_get_module_state(module)`: return the composable state if this module is managed by composable API.
`_get_module_fsdp_state(module)`: return the FSDP state if this module is managed by FSDP.

**What does this PR propose?**
1. Make `_State` out of `_composable` module so that `FullyShardedDataParallel` can inherit from it.
2. A global `_module_state_mapping: Dict[nn.Module, _State]` that keeps the mapping of all submodules (not just root module) to the state.
3. Create `_get_module_state(module)` to look up `_module_state_mapping`.
4. Create `_get_module_fsdp_state(module)` that uses `_get_module_state(module)` to get the state then verifies if the state is `_FSDPState`.



[ghstack-poisoned]

fegin committed:
Update on "[Composable API]Common _State parent class for composable and wrapper FSDP"


**Why this PR?**

For the composable APIs implementation, sometimes the internal APIs may not have the application (FSDP, DDP) root module but only the local module. One example is the state_dict/optimizer_state_dict implementation of FSDP. These APIs  are designed to start with the root module of the model. It is tricky for these APIs to tell whether a random submodule is managed by either DDP or FSDP. 

It will be useful to have APIs like:
`_get_module_state(module)`: return the composable state if this module is managed by composable API.
`_get_module_fsdp_state(module)`: return the FSDP state if this module is managed by FSDP.

**What does this PR propose?**
1. Make `_State` out of `_composable` module so that `FullyShardedDataParallel` can inherit from it.
2. A global `_module_state_mapping: Dict[nn.Module, _State]` that keeps the mapping of all submodules (not just root module) to the state.
3. Create `_get_module_state(module)` to look up `_module_state_mapping`.
4. Create `_get_module_fsdp_state(module)` that uses `_get_module_state(module)` to get the state then verifies if the state is `_FSDPState`.



[ghstack-poisoned]

fegin commented:
@pytorchbot merge

fegin committed:
Update on "[Composable API]Common _State parent class for composable and wrapper FSDP"


**Why this PR?**

For the composable APIs implementation, sometimes the internal APIs may not have the application (FSDP, DDP) root module but only the local module. One example is the state_dict/optimizer_state_dict implementation of FSDP. These APIs  are designed to start with the root module of the model. It is tricky for these APIs to tell whether a random submodule is managed by either DDP or FSDP. 

It will be useful to have APIs like:
`_get_module_state(module)`: return the composable state if this module is managed by composable API.
`_get_module_fsdp_state(module)`: return the FSDP state if this module is managed by FSDP.

**What does this PR propose?**
1. Make `_State` out of `_composable` module so that `FullyShardedDataParallel` can inherit from it.
2. A global `_module_state_mapping: Dict[nn.Module, _State]` that keeps the mapping of all submodules (not just root module) to the state.
3. Create `_get_module_state(module)` to look up `_module_state_mapping`.
4. Create `_get_module_fsdp_state(module)` that uses `_get_module_state(module)` to get the state then verifies if the state is `_FSDPState`.



[ghstack-poisoned]

fegin commented:
@pytorchbot merge
